modules/backend/models/User.php

```diff
@@ -209,4 +209,56 @@ public function unsuspend()
     {
         BackendAuth::findThrottleByUserId($this->id)->unsuspend();
     }
+
+    //
+    // Impersonation
+    //
+
+    /**
+     * Returns an array of merged permissions based on the user's individual permissions
+     * and their group permissions filtering out any permissions the impersonator doesn't
+     * have access to (if the current user is being impersonated)
+     *
+     * @return array
+     */
+    public function getMergedPermissions()
+    {
+        if (!$this->mergedPermissions) {
+            $permissions = parent::getMergedPermissions();
+
+            // If the user is being impersonated filter out any permissions the impersonator doesn't have access to already
+            if (BackendAuth::isImpersonator() && $impersonator = BackendAuth::getImpersonator() && $impersonator !== $this) {
+                foreach ($permissions as $i => $permission) {
+                    if (!$impersonator->hasAccess($permission)) {
+                        unset($permissions[$i]);
+                    }
+                }
+                $this->mergedPermissions = $permissions;
+            }
+        }
+
+        return $this->mergedPermissions;
+    }
+
+    /**
+     * Check if this user can be impersonated by the provided impersonator
+     * Super users cannot be impersonated and all users cannot be impersonated unless there is an impersonator
+     * present and the impersonator has access to `backend.impersonate_users`, and the impersonator is not the
+     * user being impersonated
+     *
+     * @param static|false $impersonator The user attempting to impersonate this user, false when not available
+     * @return boolean
+     */
+    public function canBeImpersonated($impersonator = false)
+    {
+        if (
+            $this->isSuperUser() ||
+            !$impersonator ||
```

Comment:
Instead of `false`, should this be `null`? Better describes an absence of an impersonator.

```suggestion
     * @param static|null $impersonator The user attempting to impersonate this user, `null` when not available
     * @return boolean
     */
    public function canBeImpersonated($impersonator = null)
    {
        if (
            $this->isSuperUser() ||
            is_null($impersonator) ||
```

Reply:
Might also be worth having a method existence check for the `hasAccess` method if the impersonator is provided.

Reply:
No because null will screw with Session::has()

Reply:
hasAccess() is present on the base user model instance already

Reply:
Fair enough, then the property should be type-hinted to the base user model instead. Given impersonation is supposed to be between two "user" models, we should be strict on what is allowed to be compared for personation.

Reply:
You mean instead of static?

Reply:
I mean `Winter\Storm\Auth\Models\User`, which every user model should be extended from.

Reply:
Can you submit the your idea as a suggestion?

Reply:
Sorry, I should be more clear. So I think the base `canBeImpersonated` method in `Winter\Storm\Auth\Models\User` should have that model as the type hint, which would enforce the same type-hint on the User models in the Backend module and the User plugin. If the intention is interoperability between models, it should be that the method allows any type of user.

Reply:
We'd then need to restrict the checks on the final models whether they will allow interoperability or not. In the case of the Backend user model, I would imagine that it can only be impersonated by other Backend users of a specific level of access, so you would need to do a check within its `canBeImpersonated` method that the given impersonatee is a Backend user too.

Reply:
Makes sense